internal/controller/postgrescluster/postgres.go

```diff
@@ -628,6 +635,71 @@ func (r *Reconciler) reconcilePostgresDataVolume(
+func (r *Reconciler) setVolumeSize(ctx context.Context, cluster *v1beta1.PostgresCluster,
+	pvc *corev1.PersistentVolumeClaim, instanceSpecName string) {
+	log := logging.FromContext(ctx)
+
+	// Capture the largest pgData volume size currently defined for a given instance set.
+	volumeRequestSize := pvc.Spec.Resources.Requests.Storage() //.Value()
+
+	// If the request value is greater than the set limit, use the limit and issue
+	// a warning event. A limit of 0 is ignorned.
+	if !pvc.Spec.Resources.Limits.Storage().IsZero() &&
+		pvc.Spec.Resources.Requests.Storage().Value() > pvc.Spec.Resources.Limits.Storage().Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "VolumeRequestOverLimit",
+			"pgData volume request (%v) for %s/%s is greater than set limit (%v). Limit value will be used.",
+			pvc.Spec.Resources.Requests.Storage(), cluster.Name, instanceSpecName, pvc.Spec.Resources.Limits.Storage())
+
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(pvc.Spec.Resources.Limits.Storage().Value(), resource.BinarySI),
+		}
+		// Otherwise, if the limit is not set or the feature gate is not enabled, do not autogrow.
+	} else if !pvc.Spec.Resources.Limits.Storage().IsZero() && util.DefaultMutableFeatureGate.Enabled(util.AutoGrowVolumes) {
+		for i := range cluster.Status.InstanceSets {
+			if instanceSpecName == cluster.Status.InstanceSets[i].Name {
+				for _, dpv := range cluster.Status.InstanceSets[i].DesiredPGDataVolume {
+					if dpv != "" {
+						desiredRequest, err := resource.ParseQuantity(dpv)
+						if err == nil {
+							if desiredRequest.Value() > volumeRequestSize.Value() {
+								volumeRequestSize = &desiredRequest
+							}
+						} else {
+							log.Error(err, "Unable to parse volume request: "+dpv)
+						}
+					}
+				}
+			}
+		}
+
+		// If the volume request size is greater than or equal to the limit and the
+		// limit is not zero, update the request size to the limit value.
+		// If the user manually requests a lower limit that is smaller than the current
+		// or requested volume size, it will be ignored in favor of the limit value.
+		if volumeRequestSize.Value() >= pvc.Spec.Resources.Limits.Storage().Value() {
+
+			r.Recorder.Eventf(cluster, corev1.EventTypeNormal, "VolumeLimitReached",
+				"pgData volume(s) for %s/%s are at size limit (%v).", cluster.Name,
+				instanceSpecName, pvc.Spec.Resources.Limits.Storage())
+
+			// If the volume size request is greater than the limit, issue an
+			// additional event warning.
+			if volumeRequestSize.Value() > pvc.Spec.Resources.Limits.Storage().Value() {
+				r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "DesiredVolumeAboveLimit",
+					"The desired size (%v) for the %s/%s pgData volume(s) is greater than the size limit (%v).",
+					volumeRequestSize, cluster.Name, instanceSpecName, pvc.Spec.Resources.Limits.Storage())
+			}
+
+			volumeRequestSize = pvc.Spec.Resources.Limits.Storage()
```

Comment:
:pick
```suggestion
			volumeRequestSize = volumeLimitSize
```

Reply:
See the comment above. I think I'd like to keep these values as explicitly from the pvc spec.

Reply:
Updated, see comment.